Panic on SHOW FIELD KEYS

### Bug report

__System info: InfluxDB starting, version 0.13.0~rc2, branch master, commit 590b3f9fc76904c0035ef1e9731ad1c27903ef76

__Steps to reproduce:__
SHOW FIELD KEYS  in admin interface

__Expected behavior:
```
query] 2016/05/18 10:39:45 SELECT fieldKey, fieldType FROM results."default"._fieldKeys
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x18 pc=0x64a412]

goroutine 194 [running]:
panic(0xb62300, 0xc820010070)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/influxdata/influxdb/tsdb.(*fieldKeysIterator).Next(0xc824fda180, 0xc824b40a20, 0x0, 0x0)
	/root/go/src/github.com/influxdata/influxdb/tsdb/shard.go:1019 +0x392
github.com/influxdata/influxdb/influxql.(*floatSortedMergeIterator).pop(0xc824ee0d80, 0x0, 0x0, 0x0)
	/root/go/src/github.com/influxdata/influxdb/influxql/iterator.gen.go:360 +0x190
github.com/influxdata/influxdb/influxql.(*floatSortedMergeIterator).Next(0xc824ee0d80, 0x4, 0x0, 0x0)
	/root/go/src/github.com/influxdata/influxdb/influxql/iterator.gen.go:349 +0x2d
github.com/influxdata/influxdb/influxql.(*floatInterruptIterator).Next(0xc823e6eae0, 0x7f90d07cdf40, 0x0, 0x0)
	/root/go/src/github.com/influxdata/influxdb/influxql/iterator.gen.go:750 +0xb5
github.com/influxdata/influxdb/influxql.(*floatDedupeIterator).Next(0xc823e6eb00, 0x0, 0x0, 0x0)
	/root/go/src/github.com/influxdata/influxdb/influxql/iterator.gen.go:1973 +0x5e
github.com/influxdata/influxdb/influxql.(*bufFloatIterator).Next(0xc823e6eb40, 0xc8254b9d80, 0x0, 0x0)
	/root/go/src/github.com/influxdata/influxdb/influxql/iterator.gen.go:96 +0x69
github.com/influxdata/influxdb/influxql.(*floatAuxIterator).stream(0xc82553ac30)
	/root/go/src/github.com/influxdata/influxdb/influxql/iterator.gen.go:819 +0x2b
created by github.com/influxdata/influxdb/influxql.(*floatAuxIterator).Start
	/root/go/src/github.com/influxdata/influxdb/influxql/iterator.gen.go:781 +0x35
```

__Additional info: I erased some tsm files due to compaction panic error.  If it's related you could close it

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -13,7 +13,6 @@
 - [#6519](https://github.com/influxdata/influxdb/issues/6519): Support cast syntax for selecting a specific type.
 - [#6654](https://github.com/influxdata/influxdb/pull/6654): Add new HTTP statistics to monitoring
 
-
 ### Bugfixes
 
 - [#6604](https://github.com/influxdata/influxdb/pull/6604): Remove old cluster code
@@ -24,7 +23,7 @@
 - [#6644](https://github.com/influxdata/influxdb/issues/6644): Print the query executor's stack trace on a panic to the log.
 - [#6650](https://github.com/influxdata/influxdb/issues/6650): Data race when dropping a database immediately after writing to it
 - [#6235](https://github.com/influxdata/influxdb/issues/6235): Fix measurement field panic in tsm1 engine.
-
+- [#6663](https://github.com/influxdata/influxdb/issues/6663): Fixing panic in SHOW FIELD KEYS.
 
 ## v0.13.0 [2016-05-12]
 

--- a/tsdb/shard.go
+++ b/tsdb/shard.go
@@ -486,13 +486,10 @@ func (s *Shard) FieldDimensions(sources influxql.Sources) (fields map[string]inf
 			}
 
 			// Append fields and dimensions.
-			fieldNames := mm.FieldNames()
-			if len(fieldNames) > 0 {
-				mf := s.engine.MeasurementFields(m.Name)
-				for _, name := range fieldNames {
-					if f := mf.Field(name); f != nil {
-						fields[name] = f.Type
-					}
+			mf := s.engine.MeasurementFields(m.Name)
+			if mf != nil {
+				for name, typ := range mf.FieldSet() {
+					fields[name] = typ
 				}
 			}
 			for _, key := range mm.TagKeys() {
@@ -997,7 +994,7 @@ type fieldKeysIterator struct {
 	mms Measurements // remaining measurements
 	buf struct {
 		mm     *Measurement // current measurement
-		fields []*Field     // current measurement's fields
+		fields []Field      // current measurement's fields
 	}
 }
 
@@ -1017,15 +1014,23 @@ func (itr *fieldKeysIterator) Next() (*influxql.FloatPoint, error) {
 			}
 
 			itr.buf.mm = itr.mms[0]
-			keys := itr.buf.mm.FieldNames()
-			if len(keys) > 0 {
-				// Sort the keys in alphabetical order.
+			mf := itr.sh.engine.MeasurementFields(itr.buf.mm.Name)
+			if mf != nil {
+				fset := mf.FieldSet()
+				if len(fset) == 0 {
+					itr.mms = itr.mms[1:]
+					continue
+				}
+
+				keys := make([]string, 0, len(fset))
+				for k := range fset {
+					keys = append(keys, k)
+				}
 				sort.Strings(keys)
-				// Retrieve the field for each key.
-				mf := itr.sh.engine.MeasurementFields(itr.buf.mm.Name)
-				itr.buf.fields = make([]*Field, len(keys))
+
+				itr.buf.fields = make([]Field, len(keys))
 				for i, name := range keys {
-					itr.buf.fields[i] = mf.Field(name)
+					itr.buf.fields[i] = Field{Name: name, Type: fset[name]}
 				}
 			}
 			itr.mms = itr.mms[1:]

